#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-39_noepg.dpatch from version the vdr-noepgmenu plugin 0.0.6.beta3
## http://winni.vdr-developer.org/noepgmenu/downloads/beta
##
## Thomas Günther <tom@toms-cafe.de>:
##   - adapted to VDR-1.7.8
##   - solved conflicts with other patches
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: The patch allows to disable normal epg update for specified channels.
## DP: This is useful if you get epg data of the channels from external sources.

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' vdr-1.7.19~/config.c vdr-1.7.19/config.c
--- vdr-1.7.19~/config.c	2011-06-19 20:05:17.000000000 +0200
+++ vdr-1.7.19/config.c	2011-06-19 20:08:33.000000000 +0200
@@ -329,6 +329,8 @@
   SetSystemTime = 0;
   TimeSource = 0;
   TimeTransponder = 0;
+  noEPGMode=0;
+  noEPGList=strdup("");
   MarginStart = 2;
   MarginStop = 10;
   AudioLanguages[0] = -1;
@@ -414,10 +416,17 @@
   MenuCmdPosition = 0;
 }
 
+cSetup::~cSetup()
+{
+  free(noEPGList);
+}
+
 cSetup& cSetup::operator= (const cSetup &s)
 {
   memcpy(&__BeginData__, &s.__BeginData__, (char *)&s.__EndData__ - (char *)&s.__BeginData__);
   InitialChannel = s.InitialChannel;
+  free(noEPGList);
+  noEPGList = strdup(s.noEPGList);
   return *this;
 }
 
@@ -534,6 +543,11 @@
   else if (!strcasecmp(Name, "SetSystemTime"))       SetSystemTime      = atoi(Value);
   else if (!strcasecmp(Name, "TimeSource"))          TimeSource         = cSource::FromString(Value);
   else if (!strcasecmp(Name, "TimeTransponder"))     TimeTransponder    = atoi(Value);
+  else if (!strcasecmp(Name, "noEPGMode"))           noEPGMode          = atoi(Value);
+  else if (!strcasecmp(Name, "noEPGList")) {
+    free(noEPGList);
+    noEPGList=strdup(Value ? Value : "");
+  }
   else if (!strcasecmp(Name, "MarginStart"))         MarginStart        = atoi(Value);
   else if (!strcasecmp(Name, "MarginStop"))          MarginStop         = atoi(Value);
   else if (!strcasecmp(Name, "AudioLanguages"))      return ParseLanguages(Value, AudioLanguages);
@@ -643,6 +657,8 @@
   Store("SetSystemTime",      SetSystemTime);
   Store("TimeSource",         cSource::ToString(TimeSource));
   Store("TimeTransponder",    TimeTransponder);
+  Store("noEPGMode",          noEPGMode);
+  Store("noEPGList",          noEPGList);
   Store("MarginStart",        MarginStart);
   Store("MarginStop",         MarginStop);
   StoreLanguages("AudioLanguages", AudioLanguages);
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' vdr-1.7.19~/config.h vdr-1.7.19/config.h
--- vdr-1.7.19~/config.h	2011-06-19 20:05:17.000000000 +0200
+++ vdr-1.7.19/config.h	2011-06-19 20:06:21.000000000 +0200
@@ -236,6 +236,7 @@
   int SetSystemTime;
   int TimeSource;
   int TimeTransponder;
+  int noEPGMode;
   int MarginStart, MarginStop;
   int AudioLanguages[I18N_MAX_LANGUAGES + 1];
   int DisplaySubtitles;
@@ -305,7 +306,9 @@
   int ShowRecDate, ShowRecTime, ShowRecLength, ShowProgressBar, MenuCmdPosition;
   int __EndData__;
   cString InitialChannel;
+  char *noEPGList; // pointer not to be flat-copied
   cSetup(void);
+  ~cSetup();
   cSetup& operator= (const cSetup &s);
   bool Load(const char *FileName);
   bool Save(void);
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' vdr-1.7.19~/eit.c vdr-1.7.19/eit.c
--- vdr-1.7.19~/eit.c	2011-06-19 20:05:17.000000000 +0200
+++ vdr-1.7.19/eit.c	2011-06-19 20:05:17.000000000 +0200
@@ -24,8 +24,28 @@
 class cEIT : public SI::EIT {
 public:
   cEIT(cSchedules *Schedules, int Source, u_char Tid, const u_char *Data, bool OnlyRunningStatus = false);
+
+private:
+  bool allowedEPG(tChannelID kanalID);
   };
 
+bool cEIT::allowedEPG(tChannelID kanalID) {
+  bool rc;
+
+  if (Setup.noEPGMode == 1) {
+     rc=false;
+     if (strstr(::Setup.noEPGList,kanalID.ToString())!=NULL)
+        rc=true;
+     }
+  else {
+     rc=true;
+     if (strstr(::Setup.noEPGList,kanalID.ToString())!=NULL)
+        rc=false;
+     }
+
+  return rc;
+}
+
 cEIT::cEIT(cSchedules *Schedules, int Source, u_char Tid, const u_char *Data, bool OnlyRunningStatus)
 :SI::EIT(Data, false)
 {
@@ -37,6 +57,12 @@
   if (!channel)
      return; // only collect data for known channels
 
+  // only use epg from channels not blocked by noEPG-patch
+  tChannelID kanalID;
+  kanalID=channel->GetChannelID();
+  if (!allowedEPG(kanalID))
+    return;
+
   cSchedule *pSchedule = (cSchedule *)Schedules->GetSchedule(channel, true);
 
   bool Empty = true;
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' vdr-1.7.19~/menu.c vdr-1.7.19/menu.c
--- vdr-1.7.19~/menu.c	2011-06-19 20:05:17.000000000 +0200
+++ vdr-1.7.19/menu.c	2011-06-19 20:05:17.000000000 +0200
@@ -2850,6 +2850,7 @@
 
 class cMenuSetupEPG : public cMenuSetupBase {
 private:
+  const char *noEPGModes[2];
   int originalNumLanguages;
   int numLanguages;
   void Setup(void);
@@ -2872,6 +2873,9 @@
 {
   int current = Current();
 
+  noEPGModes[0]=tr("Blacklist");
+  noEPGModes[1]=tr("Whitelist");
+
   Clear();
 
   Add(new cMenuEditIntItem( tr("Setup.EPG$EPG scan timeout (h)"),      &data.EPGScanTimeout));
@@ -2881,6 +2885,7 @@
   Add(new cMenuEditBoolItem(tr("Setup.EPG$Set system time"),           &data.SetSystemTime));
   if (data.SetSystemTime)
      Add(new cMenuEditTranItem(tr("Setup.EPG$Use time from transponder"), &data.TimeTransponder, &data.TimeSource));
+  Add(new cMenuEditStraItem(tr("Setup.EPG$Mode noEPG-Patch"),          &data.noEPGMode, 2, noEPGModes));
   // TRANSLATORS: note the plural!
   Add(new cMenuEditIntItem( tr("Setup.EPG$Preferred languages"),       &numLanguages, 0, I18nLanguages()->Size()));
   for (int i = 0; i < numLanguages; i++)
